### DISCUSSION TITLE
🎛️ Stimulus adoption discussion

### DISCUSSION THREAD

**Discussion by lb-** [wagtail/wagtail#11125] [Development]
A new discussion for the community to discuss the Stimulus adoption.

As release 5.2  will contain documented support for using Stimulus within the Wagtail admin, it would be good to have a place to collect feedback and discuss the progress of RFC78 generally.

### Links

- [Using Stimulus in Wagtail (5.2)](https://docs.wagtail.org/en/latest/extending/extending_client_side.html#extending-with-stimulus)
- [Spreadsheet roadmap](https://docs.google.com/spreadsheets/d/1LdrXlj8OeCWy3B_moYZ-ynhfZZtFVHPahG9GFoT-XBs/edit#gid=0)
- [GitHub project](https://github.com/orgs/wagtail/projects/14/views/1)
- [RFC78](https://github.com/wagtail/rfcs/pull/78)
- Past updates
  - Comment: https://github.com/wagtail/roadmap/issues/57#issuecomment-1579390531
  - Comment: https://github.com/wagtail/roadmap/issues/61#issuecomment-1670916504

**Comment by lb-** [comment]
## October 2023 Update

As it's been about a year since [RFC78 was merged](https://github.com/wagtail/rfcs/pull/78#issuecomment-1288590092), I thought it would be good to give an update.

### Wins

- 20 controllers now in the main code with 9 different contributors involved directly with this migration, many more when it comes to reviewing PRs and providing feedback on issues.
- Removal of a huge amount of inline scripts for arbitrary JavaScript, ad-hoc JS imports and jQuery usage.
- We will have officially supported documentation for using Stimulus within Wagtail in 5.2, review the RC now to give feedback. See the latest (not yet released) docs page - https://docs.wagtail.org/en/latest/extending/extending_client_side.html#extending-with-stimulus
- All of these have robust Jest unit tests in place to help us ensure they do not break in the future.
- Many of these controllers have been able to be used/adapted to replace duplicate implementations of similar functionality throughout Wagtail's JavaScript.
- A chance for many new contributors to play a part in small to medium migrations after the initial Outreachy project completed.

### State of controllers

| #   | Controller              | Identifier(s)                   | Description                                                                                          |
| --- | ----------------------- | ------------------------------- | ---------------------------------------------------------------------------------------------------- |
| 1.  | `ActionController`      | `'w-action'`                    | Supports common actions to be triggered by events such as click/reset and submit a related form.     |
| 2.  | `AutosizeController`    | `'w-autosize'`                  | Wrapper around the text area autosize NPM package.                                                   |
| 3.  | `BulkController`        | `'w-bulk'`                      | Starting implementation for bulk checkbox actions.                                                   |
| 4.  | `CloneController`       | `'w-clone'`, `'w-messages'`     | Supports event driven cloning of some element with custom content injected.                          |
| 5.  | `CountController`       | `'w-count'`                     | Allows for classes/summary of the number of found elements in a controller.                          |
| 6.  | `DialogController`      | `'w-dialog'`                    | Wrapper around a11y dialog NPM package.                                                              |
| 7.  | `DismissibleController` | `'w-dismissible'`               | Allows for some content to be closed/removed via an action.                                          |
| 8.  | `DropdownController`    | `'w-dropdown'`                  | Wrapper around tippy.js for dropdowns.                                                               |
| 9.  | `InitController`        | `'w-init'`                      | Supports adding or removing classes, dispatching events when ready or loaded.                        |
| 10. | `ProgressController`    | `'w-progress'`                  | Adds the ability to indicate that something is loaded, usually when clicked.                         |
| 11. | `RevealController`      | `'w-breadcrumbs'`, `'w-reveal'` | Supports open/close (expand/collapse) functionality on elements.                                     |
| 12. | `SkipLinkController`    | `'w-skip-link'`                 | Allows directing focus to another element, temporarily adding tabindex if needed.                    |
| 13. | `SlugController`        | `'w-slug'`                      | Allows an input's value to be cleaned, using slugify utils.                                          |
| 14. | `SubmitController`      | `'w-submit'`                    | Creates a temporary `form` element for dynamic submission of some data to an endpoint.               |
| 15. | `SwapController`        | `'w-swap'`                      | Allows async requests to be resolved and then patched into (swapped) some other location in the DOM. |
| 18. | `SyncController`        | `'w-sync'`                      | Allows one field to sync/push it's value to another field.                                           |
| 17. | `TagController`         | `'w-tag'`                       | Wrapper around the jQuery taggit library.                                                            |
| 18. | `TeleportController`    | `'w-teleport'`                  | Allows a template element to be used, it's content copied and moved to another place in the DOM.     |
| 19. | `TooltipController`     | `'w-tooltip'`                   | Wrapper around tippy.js for simple tooltip scenarios.                                                |
| 20. | `UpgradeController`     | `'w-upgrade'`                   | Specific implementation for the upgrade notification on the dashboard for admin users.               |

### Short term roadmap

1. Clean up last of the non-Stimulus attributes usage of tooltips. https://github.com/wagtail/wagtail/pull/10894
2. Replace implementations of conditional fields with new `w-cond`. https://github.com/wagtail/wagtail/issues/11045
3. Remove legacy dropdown, migrate to `w-dropdown`. https://github.com/wagtail/wagtail/pull/11084 (Draft)
4. Replace side panel & open/close panels with `w-reveal` enhancements. https://github.com/wagtail/wagtail/issues/10168
5. Migrate the page listing re-ordering (drag & drop) to a Stimulus wrapper. https://github.com/wagtail/wagtail/issues/10909
6. Remove date pickers, not strictly Stimulus now but will mean much less complex JS. https://github.com/wagtail/wagtail/issues/8056
7. Research how we can replace the current focal point chooser and the multiple file upload jQuery implementations with other libraries. https://github.com/wagtail/wagtail/issues/10947 & related issues.

### Longer term

Many of these are not yet created as a new issue but will be created when the short term backlog is reduced.

- Tabs https://github.com/wagtail/wagtail/issues/8408
- Remove our need for the additional JS library for keyboard shortcuts, blocked for now due to a Stimulus code change but it's hopefully close. https://github.com/wagtail/wagtail/issues/10167
- Dirty form checks
- `InlinePanel` / formset
- Bulk actions
- Choosers & modal workflows
- Formal support for CSP https://github.com/wagtail/roadmap/issues/61

### Enhanced features (feedback requested)

If anyone would like to see some of these features make their way to be used in a more documented/Python driven way.

Please have a look at these items and add any thoughts or `+1`s to the issues, or if you are keen to contribute, add a comment.

- Add the ability to use the sync controller for more complex cases in `FieldPanel` or a potential `SyncPanel`. https://github.com/wagtail/wagtail/issues/1063
- Add the ability for client-side dynamic conditional fields. https://github.com/wagtail/wagtail/issues/2172
- Add the ability to support re-ordering for Snippets. https://github.com/wagtail/wagtail/issues/10816
- Add the ability to restrict field values in more dynamic ways (could leverage the code within w-slug) https://github.com/wagtail/wagtail/issues/7417

**Comment by lb-** [comment]
## hotwire.io

One of the core contributors to Stimulus/Turbo just built a new site. It looks pretty good and has a great range of content for those using Stimulus. 

https://hotwire.io/

I've just raised an issue to see if we can cross reference Wagtail's support for Stimulus on this site. https://github.com/marcoroth/hotwire.io/issues/57

For anyone building custom controllers, these sections have some great resources. 

- https://hotwire.io/use-cases
- https://hotwire.io/ecosystem/components
- https://hotwire.io/ecosystem/npm

**Reply by lb-** [reply]
PR up https://github.com/marcoroth/hotwire.io/pull/59

**Reply by lb-** [reply]
Note: This got merged and is available to view live here https://hotwire.io/frameworks/wagtail

**Comment by lb-** [comment]
## Other projects using Stimulus

| #    | Website                  | Description          | Language | Repo                                            | `controllers/` folder                                                                               |
| ---- | ------------------------ | -------------------- | -------- | ----------------------------------------------- | --------------------------------------------------------------------------------------------------- |
| `43` | https://avohq.io         | CMS & CRM builder    | Ruby     | https://github.com/avo-hq/avo                   | [📁](https://github.com/avo-hq/avo/tree/main/app/javascript/js/controllers)                         |
| `41` | https://docs.part-db.de/ | Inventory management | PHP      | https://github.com/Part-DB/Part-DB-server       | [📁](https://github.com/Part-DB/Part-DB-server/tree/master/assets/controllers)                      |
| `40` | https://orchid.software/ | CMS & CRM builder    | PHP      | https://github.com/orchidsoftware/platform      | [📁](https://github.com/orchidsoftware/platform/tree/master/resources/js/controllers)               |
| `34` | https://spinacms.com/    | CMS                  | Ruby     | https://github.com/SpinaCMS/Spina               | [📁](https://github.com/SpinaCMS/Spina/tree/main/app/assets/javascripts/spina/controllers)          |
| `23` | https://bullettrain.co/  | SASS toolkit         | Ruby     | https://github.com/bullet-train-co/bullet_train | [📁](https://github.com/bullet-train-co/bullet_train/blob/main/app/javascript/controllers/index.js) |
| `20` | https://pypi.org/        | Package index        | Python   | https://github.com/pypi/warehouse               | [📁](https://github.com/pypi/warehouse/tree/main/warehouse/static/js/warehouse/controllers)         |
| `19` | https://forem.com        | Discussion forum     | Ruby     | https://github.com/forem/forem                  | [📁](https://github.com/forem/forem/tree/main/app/javascript/admin/controllers)                     |
| `12` | https://railsdevs.com/   | Blog/content site    | Ruby     | https://github.com/joemasilotti/railsdevs.com   | [📁](https://github.com/joemasilotti/railsdevs.com/tree/main/app/javascript/controllers)            |

> Sorted by amount of Stimulus controllers, only projects with 10+ controllers.

**Reply by lb-** [reply]
@thibaudcolas @laymonage  - as per our discussion the other week, you wanted to see some examples of larger projects and how they are using Stimulus. Obviously a lot of these are Ruby but here's some I have sourced from Twitter and GitHub.

All have at least 10 custom built controllers in their code and a few have over 40.

Once the projects get to around 30, they appear to start splitting into sub-folders internal sub-packages. I think we could learn a lot from these approaches, including some of the kinds of controllers they have are very similar to ours.

Let me know if there is any more digging you wanted to me to do.

**Reply by lb-** [reply]
Links for good 'reference' sets of component implementations.

* https://github.com/Sub-Xaero/stimulus-library
* https://github.com/stimulus-components
* https://github.com/excid3/tailwindcss-stimulus-components
* https://github.com/stimulus-use/stimulus-use
* https://github.com/marcoroth/formulus
* https://github.com/archonic/limestone (boilerplate Ruby package)

**Comment by lb-** [comment]
## October 2024 Update

It's now been about two years since the [Stimulus adoption RFC has been merged](https://github.com/wagtail/rfcs/pull/78#issuecomment-1288590092), there's been an incredible amount of momentum since then.

A quick summary of the wins from the last year.

- **9 new Stimulus controllers**, bringing the total to **29**.
- A significant reduction in inline scripts from **37 `<script>` elements used in active code** in Wagtail 5.2 down to **12** in Wagtail 6.3. Not all of these are direct Stimulus migrations.
- **10** new contributors have added to the `src/controllers/` folder, bringing the total to **19**, with many new contributors continuing to be involved.
- **Two new major features** have been built on top of Stimulus; concurrent editor notifications and universal filter 'drilldown' UI.
- Migration of some core and complex functionality such as unsaved changes & the preview sidebar. Meaning these now follow similar patterns for implementations, are built with TypeScript and have high levels of unit test coverage.
- Less duplication across our React and Stimulus code with being able to leverage published NPM modules easier.
- Significantly less reliance on `window.*` global variables to communicate between modules, with browser DOM events & the powerful [Stimulus outlets](https://stimulus.hotwired.dev/reference/outlets).

### Controller landscape

An overview of all our controllers, as at Wagtail 6.3.

| #     | Controller              | Identifier(s)                   | Description                                                                                          |
| ----- | ----------------------- | ------------------------------- | ---------------------------------------------------------------------------------------------------- |
| 1.    | `ActionController`      | `'w-action'`                    | Supports common actions to be triggered by events such as click/reset and submit a related form.     |
| 2.    | `AutosizeController`    | `'w-autosize'`                  | Wrapper around the text area autosize NPM package.                                                   |
| 3.🆕  | `BlockController`       | `'w-bulk'`                      | Unpacks a Telepath object and renders it on the controlled element.                                  |
| 4.    | `BulkController`        | `'w-bulk'`                      | Starting implementation for bulk checkbox actions.                                                   |
| 5.🆕  | `ClipboardController`   | `w-clipboard`                   | Adds the ability to copy contents to the clipboard.                                                  |
| 6.    | `CloneController`       | `'w-clone'`, `'w-messages'`     | Supports event driven cloning of some element with custom content injected.                          |
| 7.    | `CountController`       | `'w-count'`                     | Allows for classes/summary of the number of found elements in a controller.                          |
| 8.    | `DialogController`      | `'w-dialog'`                    | Wrapper around a11y dialog NPM package.                                                              |
| 9.    | `DismissibleController` | `'w-dismissible'`               | Allows for some content to be closed/removed via an action.                                          |
| 10.🆕 | `DrilldownController`   | `w-drilldown`                   | Adds a progressively nested set of menu options that can be 'drilled' into.                          |
| 11.   | `DropdownController`    | `'w-dropdown'`                  | Wrapper around tippy.js for dropdowns.                                                               |
| 12.   | `InitController`        | `'w-init'`                      | Supports adding or removing classes, dispatching events when ready or loaded.                        |
| 13.🆕 | `KeyboardController`    | `w-keyboard`                    | Attaches keyboard shortcuts to an element such as a button.                                          |
| 14.🆕 | `LinkController`        | `w-link`                        | Allows a link's query parameters to be dynamically updates based on `window.location`.               |
| 15.🆕 | `OrderableController`   | `w-orderable`                   | Adds drag & drop re-ordering to a set of elements.                                                   |
| 16.🆕 | `PreviewController`     | `w-preview`                     | Controls the dynamic iFrame preview for Page & Snippet previews                                      |
| 17.   | `ProgressController`    | `'w-progress'`                  | Adds the ability to indicate that something is loaded, usually when clicked.                         |
| 18.   | `RevealController`      | `'w-breadcrumbs'`, `'w-reveal'` | Supports open/close (expand/collapse) functionality on elements.                                     |
| 19.🆕 | `SessionController`     | `w-session`                     | Manages visibility of concurrent sessions with a periodic ping server event.                         |
| 20.   | `SkipLinkController`    | `'w-skip-link'`                 | Allows directing focus to another element, temporarily adding tabindex if needed.                    |
| 21.   | `SlugController`        | `'w-slug'`                      | Allows an input's value to be cleaned, using slugify utils.                                          |
| 22.   | `SubmitController`      | `'w-submit'`                    | Creates a temporary `form` element for dynamic submission of some data to an endpoint.               |
| 23.   | `SwapController`        | `'w-swap'`                      | Allows async requests to be resolved and then patched into (swapped) some other location in the DOM. |
| 24.   | `SyncController`        | `'w-sync'`                      | Allows one field to sync/push it's value to another field.                                           |
| 25.   | `TagController`         | `'w-tag'`                       | Wrapper around the jQuery taggit library.                                                            |
| 26.   | `TeleportController`    | `'w-teleport'`                  | Allows a template element to be used, it's content copied and moved to another place in the DOM.     |
| 27.   | `TooltipController`     | `'w-tooltip'`                   | Wrapper around tippy.js for simple tooltip scenarios.                                                |
| 28.🆕 | `UnsavedController`     | `w-unsaved`                     | Manages periodic checks for a `form` and dispatches events if there are unsaved changes              |
| 29.   | `UpgradeController`     | `'w-upgrade'`                   | Specific implementation for the upgrade notification on the dashboard for admin users.               |

### Code coverage

We have a really solid amount of code coverage, more work to be done but this is well above the average for our JavaScript/TypeScript.

| File                        | % Statements | % Branches | % Functions | % Lines |
| --------------------------- | ------------ | ---------- | ----------- | ------- |
| All files (average)         | 68.62        | 50.97      | 60.6        | 72.97   |
| `client/src/controllers/**` | 90.15        | 81.45      | 88.62       | 91.29   |

### Short term roadmap

- [`RulesController`](https://github.com/wagtail/wagtail/pull/11202) is close, awaiting final reviews, which will be an enabler for further jQuery removal and https://github.com/wagtail/wagtail/issues/2172 plus other 'dynamic field' requirements.
- [`TabsController](https://github.com/wagtail/wagtail/pull/11542) is being redone, hopefully with a new PR up soon that can help get this through into the main branch. This should be an enabler for https://github.com/wagtail/wagtail/issues/8408 & other Tabs improvements.
- [`CleanController` / `SlugController` improvements](https://github.com/wagtail/wagtail/pull/12379), awaiting review & core team feedback, with a potentially documented API for developers.
- We need to [add unit tests for `DrilldownController`](https://github.com/wagtail/wagtail/issues/12511), anyone can contribute to this.
- Continue to focus on inline script and jQuery items that need migration.
- Panels (side panels, field group panels, resize panels), with a starting issue raised https://github.com/wagtail/wagtail/issues/10168
- Get some clarity on our plans for expanding formsets / inline panel, potentially close or merge https://github.com/wagtail/wagtail/pull/11633

### Longer term roadmap

- Research still ongoing for date pickers, file upload, image management, tagging and others.
- Core team are working through a clearer plan for some of the other jQuery based UI such as formsets, modal workflows and choosers.
- We are well on our way to removing all inline scripts allowing our admin to be [CSP compliant](https://github.com/wagtail/wagtail/issues/1288) and we are gaining momentum on being able to remove `jQuery`.
- If there's any solid recommendations for libraries to support the roadmap that you or your teams have used in production, please add a comment.

<details>

<summary><strong>Personal learnings</strong></summary>

Some learnings from the last year-ish. My (LB) opinion only, not reflective of the core team's opinions.

- Bugs from migrations. This is probably unavoidable but frustrating when we do our best to check every edge case and find or introduce regression bugs. My hope is that at least now we have unit tests to help us fix things and know their fixed for future.
- Undocumented `window.*` globals / approaches used by developers. Some surprising cases where we have bugs raised for undocumented items that get removed or changed causing issues, I think it's a reminder that exposing anything globally means someone will find a way to use it.
- Clarity on future migration path within the core team. Recently we have found there's a bit of missing clarity within the core team about some of the more complex components and how to get them migrated. This is probably normal for any change of this scale.
- Reviews. The hardest and slowest part is doing or getting reviews from the core team (myself included, when reviewing other code), it can take 3-6 months for some items to be reviewed, hopefully as we all learn Stimulus more this will become faster but it's also the nature of a complex and mostly volunteer based open source project.
- We have some very smart people on the core team. I am so thankful to be part of this core team, my ideas are challenged (in a good way) and it's awesome to see TypeScript devs way smarter than me write better or improved versions of what was started.

</details>

**Reply by vossisboss** [reply]
Thank you so much for this fantastic update LB!

**Reply by tomdyson** [reply]
Yes, huge kudos and thanks to @lb- for making this happen. We're very lucky to have you on the team.
